lib/vdc/QuadTreeRectangleP.cpp

```diff
@@ -0,0 +1,288 @@
+            float bottom = (float)coords[1];
+
+            for (int j = 1; j < nodes.size(); j++) {
+                grid->GetUserCoordinates(nodes[j], coords);
+
+                if (coords[0] < left) left = (float)coords[0];
+                if (coords[0] > right) right = (float)coords[0];
+                if (coords[1] < top) top = (float)coords[1];
+                if (coords[1] > bottom) bottom = (float)coords[1];
+            }
+
+            // Figure out which subtree(s) contain the rectangle. In general,
+            // a rectangle can span multiple subtrees, in which case it
+            // will be inserted into both.
+            //
+            float bin_width = (_right - _left) / ((float)_qtrs.size());
+            float binLeft = _left;
+            for (int j = 0; j < _qtrs.size(); j++) {
+                float binRight = binLeft + bin_width;
+
+                if (j == _qtrs.size() - 1) binRight = _right;
+
+                if ((left <= binRight) && (right >= binLeft)) {
+                    class QuadTreeRectangle<float, pType>::rectangle_t r(left, top, right, bottom);
+
+                    pType p = {(uint32_t)cell[0], (uint32_t)cell[1]};
+#pragma omp critical
+                    {
+                        parRectangles[j].push_back(r);
+                        parPayloads[j].push_back(p);
+                    }
+                }
+
+                binLeft = binRight;
+            }
+        }
+    }
+
+    // At this point parRectangles and parPayloads contain a vector of
+    // rectangles and their payloads that intersect each of the subtrees that
+    // make up the tree. Since there is no shared data between each of
+    // the subtrees we can populate each subtree in parallel without any need
+    // for synchronization (mutex, etc.)
+    //
+    bool status = true;
+#pragma omp parallel
+#pragma omp for
+    for (int i = 0; i < _qtrs.size(); i++) {
+        for (size_t j = 0; j < parRectangles[i].size(); j++) { status &= _qtrs[i]->Insert(parRectangles[i][j], parPayloads[i][j]); }
+    }
+
+    return (status);
+}
+
+void QuadTreeRectangleP::GetPayloadContained(float x, float y, std::vector<Size_tArr3> &payloads) const
+{
+    payloads.clear();
+
+    int   bin = 0;
+    float bin_width = ((float)_right - (float)_left) / ((float)_qtrs.size());
+    float binLeft = _left;
+    for (int i = 0; i < _qtrs.size(); i++) {
+        float binRight = binLeft + bin_width;
+        if (i == _qtrs.size() - 1) binRight = _right;
+
+        if (x >= binLeft && x <= binRight) {
+            bin = i;
+            break;
+        }
+        binLeft = binRight;
+    }
+
+    std::vector<pType> p;
+    _qtrs[bin]->GetPayloadContained(x, y, p);
+
+    for (auto itr = p.begin(); itr != p.end(); ++itr) { payloads.push_back(Size_tArr3{(*itr)[0], (*itr)[1], 0}); }
+}
+
+void QuadTreeRectangleP::GetStats(std::vector<size_t> &payload_histo, std::vector<size_t> &level_histo) const
+{
+    payload_histo.clear();
+    level_histo.clear();
```

Comment:
This implementation seems incomplete and this function is never used.

Reply:
Fixed.